### ISSUE TITLE
Provide reverse-requires search in zypper

### ISSUE THREAD

**Issue by bzeller** [openSUSE/zypper#214]
Just as apt has a command called "apt-rdepends" that you can use to find all packages depending on a given package, we should have a option in zypper search (maybe reverse-requires) that allows the user to find all packages that require a given package. 

For example:
zypper se --reverse-requires libqt5core5  

Also we should have a switch wether to expand packages (i.e. 'FOO' as 'literally FOO' or 'everything FOO provides' in case it's a package)

**Comment by linuunix** [comment]
Yes, please

**Comment by mlschroe** [comment]
If you want to match against a given *package* (and not a dependency), you should not use `pool_whatmatchesdep` but `pool_whatmatchessolvable` (or `selection_make_matchsolvablelist` if you want to match against a list of packages).

**Comment by bzeller** [comment]
> If you want to match against a given _package_ (and not a dependency), you should not use `pool_whatmatchesdep` but `pool_whatmatchessolvable` (or `selection_make_matchsolvablelist` if you want to match against a list of packages).

What the code does is look for a package to extract all its "provides:" then looks for packages that require one of the provides. Would I get the same from pool_whatmatchessolvable too?

**Comment by linuunix** [comment]
While at it, please add a short command like "rr" as abbreviation for reverse-requires, other commands have such short commands too. (if, se, in,..).    
If it would make sense or is doable is another topic, but it helps to remember the commands. Especially for people new to zypper or those who use GUI most of time.

**Comment by mlandres** [comment]
Apart from that, `--reverse-requires` is an option to `search` and no command.

**Comment by mlschroe** [comment]
Regarding looking at the provides and then searching for packages that match the provides: that doesn't really work with rich dependencies. If a package has `Requires: A unless B`, you can't do that correctly without considering the package provides as a whole.

(That's actually why pool_whatmatchessolvavble was added, because dnf also did it wrong and a bug was opened against it.)

Regarding the name: `--reverse-requires` is pretty horrible, what does the `reverse` even mean in this context? We already have `--requires` which is acrually `--whatrequires`. I'd prefer something like `--requires-pkg`.

Regarding `this currently only works for installable packages`. I think libzypp only has installable packages in the pool.

**Comment by mlandres** [comment]
Maybe we have different understanding of `installable`.

**Comment by mlschroe** [comment]
Installable in the libsolv sense means that the package architecture is allowed on the system and the package is not disabled (libzypp does not do this). See `pool_installable()` in repo.h.

**Comment by bzeller** [comment]
> Regarding looking at the provides and then searching for packages that match the provides: that doesn't really work with rich dependencies. If a package has `Requires: A unless B`, you can't do that correctly without considering the package provides as a whole.

Ok then we should not do it the wrong way :D.

> Regarding the name: `--reverse-requires` is pretty horrible, what does the `reverse` even mean in this context? We already have `--requires` which is acrually `--whatrequires`. I'd prefer something like `--requires-pkg`.

Naming is not fixed yet, it was just meant as a temporary switch for now. We take suggestions.

One thing missing though, zypper se has the --verbose switch which shows a list of which depenencies have matched in the search. Any chance we can get that data from libsolv or if not already possible the API can be extended to get to the information?

**Comment by darix** [comment]
imho ... use something like "zypper what-dependson" or at least have it as an alias. similar to "zypper what-provides"

**Comment by bzeller** [comment]
> imho ... use something like "zypper what-dependson" or at least have it as an alias. similar to "zypper what-provides"

We decided to add switches to the normal search command, for now. Zypper already has many many commands adding new ones will just make it worse.

**Comment by brianjmurrell** [comment]
What version of opensuse/leap should this be available in?

**Comment by mlandres** [comment]
> What version of opensuse/leap should this be available in?

zypper-1.14.33 (1.14.*)  should go to TW and Leap15*
```
- Provide reverse search in zypper (fixes #214)
  This patch adds a new set of switches to zypper to support
  searching reverse dependencies for a package or a set of packages.
...
- version 1.14.33
```
